### PR TITLE
Fellum

### PR DESCRIPTION
**Pull request by Fellum**
Disassembler finished

### CODE REVIEW

File: src/asm/generate_commands_dis.c [hunk 1]
```diff
@@ -1,22 +1,117 @@
...
+	{
+		if (op->paramtypes[cur_par] == T_REG)
+		{
+			cmd->params[cur_par].data = read_int8_mem(exc, cr, cap);
+			cmd->params[cur_par].type = T_REG;
+		}
+		else if (op->paramtypes[cur_par] == T_IND)
+		{
+			cmd->params[cur_par].data = read_int8_mem(exc, cr, cap);
+			cmd->params[cur_par].type = T_IND;
```
Comment (TrueBadur) [inline src/asm/generate_commands_dis.c]: Feels like redundant code. may be shorten for two ifs and one initiation of cmd->params[].type

File: src/asm/generate_commands_dis.c [hunk 2]
```diff
@@ -1,22 +1,117 @@
...
+		tmp = get_par_type(cmd->par_types, cur_par);
+		if (tmp == 1 && T_REG & op->paramtypes[cur_par])
+		{
+			cmd->params[cur_par].data = read_int8_mem(exc, cr, cap);
+			cmd->params[cur_par].type = T_REG;
+		}
+		else if (tmp == 3 && T_IND & op->paramtypes[cur_par])
+		{
+			cmd->params[cur_par].data = read_int16_mem(exc, cr, cap);
+			cmd->params[cur_par].type = T_IND;
```
Comment (TrueBadur) [inline src/asm/generate_commands_dis.c]: The same as previos

File: src/asm/readers_dis.c [hunk 3]
```diff
@@ -37,4 +47,41 @@ void		read_arr(int fd, void *target, size_t size)
 		perror(strerror(errno));
 		exit(1);
 	}
-}
\ No newline at end of file
+}
+
+int			read_int32_mem(char *mem, size_t *ind, size_t cap)
```
Comment (TrueBadur) [inline src/asm/readers_dis.c]: May be rewritten as more general function

File: src/asm/writer_dis.c [hunk 4]
```diff
@@ -1,8 +1,73 @@
...
+void			write_commands_dis(t_headdata *data, t_list *cmd_lst,
+									char *ofname)
+{
+	int fd;
+
+	if ((fd = (open(ofname,
+			O_CREAT | O_WRONLY | O_TRUNC, S_IRWXU | S_IRWXG | S_IRWXO))) == -1)
+	{
+		ft_fdprintf(STDERR_FILENO, "Cannot create file(((\n");
+		exit(2);
```
Comment (TrueBadur) [inline src/asm/writer_dis.c]: And what about freeing everything?

File: src/asm/disassembler.c [hunk 5]
```diff
@@ -39,6 +39,19 @@ void	*check_header(int fd, t_headdata *headdata)
 	return (exec_code);
 }
 
+void	free_cmd_list(t_list_node *lst)
```
Comment (TrueBadur) [inline src/asm/disassembler.c]: May be you should place such function into your lib?

File: include/asm/disassembler.h [hunk 6]
```diff
@@ -20,31 +20,39 @@
...
 # define WRONG_FILE_FORMAT_ERR "Wrong file format!"
-# define SIZE_MISMATCH_ERR "Size mismatch!"
+# define INVALID_PAR_TYPE_ERR "Invalid parameter type!"
+# define INVALID_CMD_TYPE_ERR "Invalid command type!"
 
 typedef struct		s_headdata
 {
-	short			filler2;
 	char			champ_comment[COMMENT_LENGTH];
 	uint32_t		size;
```
Comment (TrueBadur) [inline include/asm/disassembler.h]: it's better to place uint32_t at first place

File: src/asm/generate_commands_dis.c (hunk 1 above)
Comment (Fellum) [reply src/asm/generate_commands_dis.c]: Not in this case - each if corresponds to each variable type, seems correct for me.

File: src/asm/generate_commands_dis.c (hunk 2 above)
Comment (Fellum) [reply src/asm/generate_commands_dis.c]: Agree with you)

File: src/asm/writer_dis.c (hunk 4 above)
Comment (Fellum) [reply src/asm/writer_dis.c]: Why should I free anything before exit?

File: include/asm/disassembler.h (hunk 6 above)
Comment (Fellum) [reply include/asm/disassembler.h]: Does it change anythin? 2, 1, 4 or 4, 2, 1 - looks the same for me.

**Pull request merged**